### PR TITLE
Twitter tool

### PR DESCRIPTION
**Pull request by Tarraann**
### Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs update

### Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have read the contributing guide and my code conforms to the guidelines.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have added the required tests.

### CODE REVIEW

**Comment by neelayan7** [comment]
Tests are failing, please check @Tarraann

**Review by DMTarmey — APPROVED** [review]
Yes

File: superagi/tools/twitter/twitter_toolkit.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+from abc import ABC
+from superagi.tools.base_tool import BaseToolkit, BaseTool
+from typing import Type, List
+from superagi.tools.twitter.send_tweets import SendTweetsTool
+
+
+class TwitterToolKit(BaseToolkit, ABC):
```
Comment (TransformerOptimus) [inline superagi/tools/twitter/twitter_toolkit.py:7]: rename to Toolkit

File: superagi/tools/twitter/send_tweets.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            headers = {'Authorization': 'application/octet-stream'}
+            response = requests.post(upload_endpoint, headers=headers,
+                             data={'media_data': b64_image},
+                             auth=oauth)
+            ids = json.loads(response.text)['media_id']
+            media_ids.append(str(ids))
+
+        return media_ids
+
+    def get_file_path(self, file_name):
```
Comment (TransformerOptimus) [inline superagi/tools/twitter/send_tweets.py:61]: use get_agent_resource_path in resource helper. Duplicate logic here.

Comment (TransformerOptimus) [reply superagi/tools/twitter/send_tweets.py:61]: Just read the output directory for now. We may not need input for now

File: superagi/tools/twitter/send_tweets.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    def _execute(self, is_media: bool, tweet_text: str = 'None', media_num: int = 0, media_files: list = []):
+        toolkit_id = self.toolkit_config.toolkit_id
+        creds = TwitterTokens().get_twitter_creds(toolkit_id)
+        params = {}
+        if is_media:
+            media_ids = self.get_media_ids(media_files, creds)
+            params["media"] = {"media_ids": media_ids}
+        if tweet_text is not None:
+            params["text"] = tweet_text
+        tweet_response = self.send_tweets(params, creds)
```
Comment (TransformerOptimus) [inline superagi/tools/twitter/send_tweets.py:34]: make a class for TwitterInput text, media

File: superagi/helper/twitter_tokens.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        api_key_secret = ""
+        for credentials in twitter_creds:
+            credentials = credentials.__dict__
+            if credentials["key"] == "TWITTER_API_KEY":
+                api_key = credentials["value"]
+            if credentials["key"] == "TWITTER_API_SECRET":
+                api_key_secret = credentials["value"]
+        creds["api_key"] = api_key
+        creds["api_key_secret"] = api_key_secret
+        return creds
```
Comment (TransformerOptimus) [inline superagi/helper/twitter_tokens.py:93]: return TwitterCred object from here.

File: superagi/helper/twitter_tokens.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            root_dir = root_dir if root_dir.endswith("/") else root_dir + "/"
+            file_path = root_dir + file_name
+        else:
+            file_path = os.getcwd() + "/" + file_name
+        if os.path.exists(file_path):
+            with open(file_path,'rb') as file:
+                creds = pickle.load(file)
+            if isinstance(creds, str):
+                creds = json.loads(creds)
+        engine = connect_db()
```
Comment (TransformerOptimus) [inline superagi/helper/twitter_tokens.py:79]: We need better way to call session.

Comment (luciferlinx101) [reply superagi/helper/twitter_tokens.py:79]: @Tarraann we have session in superagi/jobs/agent_executor.py DBToolkitConfiguration so self.session should work as you do for self.toolkit in other tools

File: superagi/helper/twitter_helper.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+            headers = {'Authorization': 'application/octet-stream'}
+            response = requests.post(upload_endpoint, headers=headers,
+                             data={'media_data': b64_image},
+                             auth=oauth)
+            ids = json.loads(response.text)['media_id']
+            media_ids.append(str(ids))
+        return media_ids
+
+    def get_file_path(self, file_name, agent_id):
+        final_path = ResourceHelper().get_agent_resource_path(file_name, agent_id)
```
Comment (TransformerOptimus) [inline superagi/helper/twitter_helper.py:31]: user ResourceHelper.get_agent_resource_path this way. it is a static method.

Comment (TransformerOptimus) [reply superagi/helper/twitter_helper.py:31]: use _get_file_path instead this is pattern for private methods.

File: superagi/models/oauth_tokens.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+        Returns:
+            str: String representation of the OauthTokens object.
+        """
+
+        return f"Tokens(id={self.id}, user_id={self.user_id}, toolkit_id={self.toolkit_id}, key={self.key}, value={self.value})"
+    
+    @staticmethod
+    def add_or_update(session: Session, toolkit_id: int, user_id: int, key: str, value: Text = None):
+        print("//////////////////////////////////////////////////")
```
Comment (TransformerOptimus) [inline superagi/models/oauth_tokens.py:42]: remove prints

Comment (TransformerOptimus) [reply superagi/models/oauth_tokens.py:42]: make it a cls method

File: migrations/versions/cc1dcc508611_table_for_oauth_tokens.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+# revision identifiers, used by Alembic.
+revision = 'cc1dcc508611'
+down_revision = '7a3e336c0fba'
+branch_labels = None
+depends_on = None
+
+
+def upgrade() -> None:
+    # ### commands auto generated by Alembic - please adjust! ###
```
Comment (TransformerOptimus) [inline migrations/versions/cc1dcc508611_table_for_oauth_tokens.py:20]: rename the file to create_oauth_tokens

File: superagi/helper/twitter_tokens.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+class Creds:
+
+    def __init__(self,api_key, api_key_secret, oauth_token, oauth_token_secret):
+        self.api_key = api_key
+        self.api_key_secret = api_key_secret
+        self.oauth_token = oauth_token
+        self.oauth_token_secret = oauth_token_secret
+
+class TwitterTokens:
```
Comment (TransformerOptimus) [inline superagi/helper/twitter_tokens.py:24]: this class does not have unit tests.

**Review by TransformerOptimus — APPROVED** [review]

File: main.py [hunk 10]
```diff
@@ -411,6 +428,40 @@ def get_google_calendar_tool_configs(toolkit_id: int):
         "client_id": google_calendar_config.value
     }
 
+@app.get("/twitter/get_twitter_creds/toolkit_id/{toolkit_id}")
+def get_twitter_tool_configs(toolkit_id: int):
```
Comment (TransformerOptimus) [inline main.py:432]: can you move this to twitter_auth.rb in controller.

**Pull request merged**